src/machine/machine_rp2040_pwm.go

```diff
@@ -0,0 +1,494 @@
+	return (pwm.slice[slicenum].csr.Get() & rp.PWM_CH0_CSR_PH_CORRECT_Msk) >> rp.PWM_CH0_CSR_PH_CORRECT_Pos
+}
+
+//go:inline
+func (pwm *pwms) getClockDiv(slicenum uint8) (Int, frac uint32) {
+	div := pwm.slice[slicenum].div.Get()
+	return (div & rp.PWM_CH0_DIV_INT_Msk) >> rp.PWM_CH0_DIV_INT_Pos, (div & rp.PWM_CH0_DIV_FRAC_Msk) >> rp.PWM_CH0_DIV_FRAC_Pos
+}
+
+// div: integer value to reduce counting rate by. Must be greater than or equal to 1.
+//
+// cc: counter compare level. Contains 2 channel levels. The 16 LSBs are Channel A's level (Duty Cycle)
+// and the 16 MSBs are Channel B's level.
+//
+// top: Wrap. Highest number counter will reach before wrapping over. usually 0xffff.
+//
+// csr: Clock mode. PWM_CH0_CSR_DIVMODE_xxx registers have 4 possible modes, of which Free-running is used.
+// csr contains output polarity bit at PWM_CH0_CSR_x_INV where x is the channel.
+// csr contains phase correction bit at PWM_CH0_CSR_PH_CORRECT_Msk.
+// csr contains PWM enable bit at PWM_CH0_CSR_EN. If not enabled PWM will not be active.
+//
+// ctr: PWM counter value.
+type pwmGroup []struct {
+	csr, cc, ctr, div, top *volatile.Register32
+}
+
+//go:inline
+func (pwm *pwms) initSlice() {
+	pwm.slice = make(pwmGroup, 8)
+	s := pwm.slice
+	hw := pwm.hw
+	// PWM0
+	s[0].cc = &hw.CH0_CC
+	s[0].csr = &hw.CH0_CSR
+	s[0].ctr = &hw.CH0_CTR
+	s[0].div = &hw.CH0_DIV
+	s[0].top = &hw.CH0_TOP
+
+	// PWM1
+	s[1].cc = &hw.CH1_CC
+	s[1].csr = &hw.CH1_CSR
+	s[1].ctr = &hw.CH1_CTR
+	s[1].div = &hw.CH1_DIV
+	s[1].top = &hw.CH1_TOP
+
+	// PWM2
+	s[2].cc = &hw.CH2_CC
+	s[2].csr = &hw.CH2_CSR
+	s[2].ctr = &hw.CH2_CTR
+	s[2].div = &hw.CH2_DIV
+	s[2].top = &hw.CH2_TOP
+
+	// PWM3
+	s[3].cc = &hw.CH3_CC
+	s[3].csr = &hw.CH3_CSR
+	s[3].ctr = &hw.CH3_CTR
+	s[3].div = &hw.CH3_DIV
+	s[3].top = &hw.CH3_TOP
+
+	// PWM4
+	s[4].cc = &hw.CH4_CC
+	s[4].csr = &hw.CH4_CSR
+	s[4].ctr = &hw.CH4_CTR
+	s[4].div = &hw.CH4_DIV
+	s[4].top = &hw.CH4_TOP
+
+	// PWM5
+	s[5].cc = &hw.CH5_CC
+	s[5].csr = &hw.CH5_CSR
+	s[5].ctr = &hw.CH5_CTR
+	s[5].div = &hw.CH5_DIV
+	s[5].top = &hw.CH5_TOP
+
+	// PWM6
+	s[6].cc = &hw.CH6_CC
+	s[6].csr = &hw.CH6_CSR
+	s[6].ctr = &hw.CH6_CTR
+	s[6].div = &hw.CH6_DIV
+	s[6].top = &hw.CH6_TOP
+
+	// PWM7
+	s[7].cc = &hw.CH7_CC
+	s[7].csr = &hw.CH7_CSR
+	s[7].ctr = &hw.CH7_CTR
+	s[7].div = &hw.CH7_DIV
+	s[7].top = &hw.CH7_TOP
+}
+
+// pwmGPIOToSlice Determine the PWM channel that is attached to the specified GPIO.
+// gpio must be less than 30. Returns the PWM slice number that controls the specified GPIO.
+//go:inline
+func pwmGPIOToSlice(gpio Pin) (slicenum uint8) {
+	return (uint8(gpio) >> 1) & 7
+}
+
+// Determine the PWM channel that is attached to the specified GPIO.
+// Each slice 0 to 7 has two channels, A and B.
+//go:inline
+func pwmGPIOToChannel(gpio Pin) (channel uint8) {
+	return uint8(gpio) & 1
+}
+
+// Returns the period of a clock cycle for the raspberry pi pico in nanoseconds.
+//go:inline
```

Comment:
Please don't use `//go:inline` unless you have measured this gives a performance or size improvement - and even then only do it when needed. In this case, I doubt it makes a difference (the compiler will likely already inline the function).